### PR TITLE
[clangd] Consider expression statements in ExtractVariable tweak

### PR DESCRIPTION
**Pull request by ckandeler**
For instance:
  int func();
  int main()
  {
    func(); // => auto placeholder = func();
  }

### CODE REVIEW

**Comment by HighCommander4** [comment]
@ckandeler Could you elaborate on the motivation for this change?

In my mind, the value proposition of the "extract variable" refactoring is that it saves you the work of moving the expression from one location to another, and typing the name of the variable twice (once at the declaration and once at the use).

In the expression-statement case, the expression isn't moving to a new location, and the variable name is only mentioned in one place. In essence, the refactoring amounts to prepending `auto placeholder = ` to the expression. Given that you're going to be (very likely) renaming `placeholder` anyways, does this really add anything meaningful over just typing `auto VariableNameYouWant = ` "manually"?

**Comment by ckandeler** [comment]
> In the expression-statement case, the expression isn't moving to a new location, and the variable name is only mentioned in 
> one place. In essence, the refactoring amounts to prepending `auto placeholder = ` to the expression. Given that you're 
> going to be (very likely) renaming `placeholder` anyways, does this really add anything meaningful over just typing `auto 
> VariableNameYouWant = ` "manually"?

My thinking was that because the expression is assignable and it's not clearly pointless to do so, the tweak should be available (plus it still saves some keystrokes). My concrete motivation was due to user requests on the client side, e.g.:
https://bugreports.qt.io/browse/QTCREATORBUG-9363
https://bugreports.qt.io/browse/QTCREATORBUG-9578
Maybe some developers have the habit of writing just the rhs of an init statement and then let tooling insert the lhs? It's possible, I suppose. E.g. if the return type is complex and you want to have it spelt out in the declaration (as per the project's coding conventions), then you could apply this tweak and then the ExpandDeducedType one.
In general, I'm not a big fan of going out of my way to prevent users from doing something. And once you start deliberating about degrees of usefulness, you quickly run into problems deciding where to draw the line.

File: clang-tools-extra/clangd/refactor/tweaks/ExtractVariable.cpp [hunk 1]
```diff
@@ -599,10 +607,22 @@ Expected<Tweak::Effect> ExtractVariable::apply(const Selection &Inputs) {
...
+  const SelectionTree::Node &OuterImplicit =
+      Target->getExprNode()->outerImplicit();
+  assert(OuterImplicit.Parent);
+  bool IsStmtExpr = llvm::isa_and_nonnull<CompoundStmt>(
+      OuterImplicit.Parent->ASTNode.get<Stmt>());
+
   // insert new variable declaration
-  if (auto Err = Result.add(Target->insertDeclaration(VarName, Range)))
+  if (auto Err =
+          Result.add(Target->insertDeclaration(VarName, Range, !IsStmtExpr)))
```
Comment (HighCommander4) [inline clang-tools-extra/clangd/refactor/tweaks/ExtractVariable.cpp:619]: The reason for needing `AddSemicolon: false` here for expression statements is a bit subtle:

 * The source range of an expression statement does not include the semicolon, so the text we're **inserting** as the initializer will not include the semicolon
 * However, the text we're **removing** below does not include the semicolon either (i.e. it's left behind), so we don't want to add another one

Could you add a comment clarifying this a bit?

File: clang-tools-extra/clangd/unittests/tweaks/ExtractVariableTests.cpp [hunk 2]
```diff
@@ -152,7 +152,7 @@ TEST_F(ExtractVariableTest, Test) {
       a = [[b]];
       a = [[xyz()]];
       // statement expression
-      [[xyz()]];
+      [[v()]];
```
Comment (HighCommander4) [inline clang-tools-extra/clangd/unittests/tweaks/ExtractVariableTests.cpp:155]: please update the comment above to "expression statement of type void"

File: clang-tools-extra/clangd/refactor/tweaks/ExtractVariable.cpp [hunk 3]
```diff
@@ -423,8 +427,6 @@ bool childExprIsStmt(const Stmt *Outer, const Expr *Inner) {
   if (!Outer || !Inner)
     return false;
   // Exclude the most common places where an expr can appear but be unused.
-  if (llvm::isa<CompoundStmt>(Outer))
```
Comment (HighCommander4) [inline clang-tools-extra/clangd/refactor/tweaks/ExtractVariable.cpp:426]: We should revise the name of this function from `childExprIsStmt`, which is no longer accurate, to something like `childExprIsDisallowedStmt`, and update the [comment at the call site](https://searchfox.org/llvm/rev/7831c5e4831963425279db7b1c4d23cfa9da9674/clang-tools-extra/clangd/refactor/tweaks/ExtractVariable.cpp#479-480) to talk about what kind of statements we allow extracting and what kind we don't.

**Review by HighCommander4 — CHANGES_REQUESTED** [review]
Ok, fair enough.

**Comment by ckandeler** [comment]
Rebased.

File: clang-tools-extra/clangd/refactor/tweaks/ExtractVariable.cpp (hunk 3 above)
Comment (ckandeler) [reply clang-tools-extra/clangd/refactor/tweaks/ExtractVariable.cpp:426]: I made the comment rather generic, as I found it difficult to summarize what the function does without essentially repeating the code.

**Comment by ckandeler** [comment]
Addressed review comments.

**Review by HighCommander4 — APPROVED** [review]
Thanks, LGTM

**Pull request merged**